### PR TITLE
Added UserProfile page

### LINKED ISSUE #1: User profile
As a new user, I want to create a user account to store my information and to be able to use the application.

### Acceptance Criteria

- [ ] **AC1 — Successful registration**
  Given an unauthenticated visitor is on the sign-up page
  When they submit a valid display name, (age), email, password, country of living and main currency
  Then an account is created and persisted
  And they are logged in and redirected to their groups overview

- [ ] **AC2 — Required fields**
  Given the visitor is on the sign-up page
  When they submit with any of the credentials empty
  Then the account is not created
  And an inline error is shown on each empty field

- [ ] **AC3 — Invalid email format**
  Given the visitor is on the sign-up page
  When they submit an email that is not in `name@domain.tld` format
  Then the account is not created
  And an error states that a valid email address is required

- [ ] **AC4 — Duplicate email**
  Given an account already exists with the email `x@y.com`
  When a visitor submits the sign-up form with `x@y.com`
  Then no second account is created
  And an error states that the email is already registered

- [ ] **AC5 — Password rules**
  Given the visitor is on the sign-up page
  When they submit a password shorter than 8 characters
  Then the account is not created
  And an error states the minimum length requirement

- [ ] **AC6 — Country of living selection**
  Given the visitor is on the sign-up page
  When they open the country of living field
  Then a scrollable list of countries is shown
  And they must select one from the list to proceed

- [ ] **AC7 — Currency suggested from country**
  Given the visitor has selected a country of living
  When the main currency field is displayed
  Then it is pre-filled with the currency matching that country
  And the visitor can change it to a different currency before submitting

- [ ] **AC8 — Credential storage (non-functional)**
  The system must store passwords only as a salted hash (BCrypt); the plaintext password must never be persisted or returned in any API response.

### PR DESCRIPTION
**Pull request by LouisCao1029**
Add User class to store account creation inputs.
Add UserController class to manage requests from frontend
Add UserRepository to communicate with database (MySQL)
Add UserService that controls logic
Added UserController class to connect backend to frontend
Added Unit tests for UserProfile

NOTE: 
Docker Desktop must be installed to run DB tests.

Note: Generative AI was used to help in generating frontend, test cases and debugging. All generated files were manually reviewed and locally tested multiple times.

### CODE REVIEW

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/model/User.java
Comment (AdRo-Coder) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/model/User.java:1]: Currently all the attributes would be nullable. In my view, all of them are must have and therefor shouldn't be nullable.  The mail is supposed to be unique. The configuration would allow to have duplicates. 
Add following annotations:
@Column(nullable = false) : to prevent null values
@Column(nullable = false, unique = true) : to ensure a value is unique and not null

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/model/User.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
+package nz.ac.auckland.se310.fairshare.model;
+
+import jakarta.persistence.*;
+
+@Entity
+@Table(name = "users")
```
Comment (AdRo-Coder) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/model/User.java:6]: I would use singulars for the table names and not plurals, but up for discussion.

**Review by AdRo-Coder — CHANGES_REQUESTED** [review]
Great work. Sharpening the rules for the db will prevent future problems. The naming is an optional feedback

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/AppConfig.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        source.registerCorsConfiguration("/**", configuration);
+
+        return source;
+    }
+  @Bean
+  public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+    http.cors(cors -> {})
+        .csrf(csrf -> csrf.disable())
+        .authorizeHttpRequests(
+            auth ->
```
Comment (Shio-y) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/AppConfig.java:54]: Could you add some comments here to explain what is happening?

File: backend/src/main/resources/db/migration/V1__create_users_table.sql [hunk 3]
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE users (
+    id BIGINT AUTO_INCREMENT PRIMARY KEY,
+    username VARCHAR(255),
+    password VARCHAR(255),
+    email VARCHAR(255),
+    country VARCHAR(255),
```
Comment (Shio-y) [inline backend/src/main/resources/db/migration/V1__create_users_table.sql:6]: currency and country should be enums instead of VARCHAR

File: backend/src/test/java/nz/ac/auckland/se310/fairshare/FairShareApplicationTests.java
Comment (Shio-y) [inline backend/src/test/java/nz/ac/auckland/se310/fairshare/FairShareApplicationTests.java:1]: should add a test assertion or at least a comment explaining why it is empty.

File: backend/src/test/java/nz/ac/auckland/se310/fairshare/TestSecurityConfig.java
Comment (Shio-y) [inline backend/src/test/java/nz/ac/auckland/se310/fairshare/TestSecurityConfig.java:1]: would also like some brief comments explaining what this does

File: backend/src/test/java/nz/ac/auckland/se310/fairshare/UserTests.java
Comment (Shio-y) [inline backend/src/test/java/nz/ac/auckland/se310/fairshare/UserTests.java:1]: tests involving multiple assertions should use assertAll() in order to account for multiple assertion failures.

**Review by Shio-y — CHANGES_REQUESTED** [review]
Overall pretty good. Just a few small changes and some comments in certain unclear methods and it should be good to merge.

**Review by AdRo-Coder — APPROVED** [review]
Thanks for the changes. Good to me now.

**Review by Shio-y — APPROVED** [review]
Great job. Should be good to merge now.

**Comment by LouisCao1029** [comment]
Addressed review comments, changes now additionally include:
annotated variables in user to be not nullable
added comments to explain the config files
refactored tests to use assertAll()
